### PR TITLE
add vscodeWithConfiguration, vscodeExts2nix, vscodeEnv

### PR DESCRIPTION
**Pull request by countoren**
move mktplcExtRefToFetchArgs to file in order to be shared with the new derivations(privately)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ X ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ X ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ X ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ X ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @jonringer @zimbatm 
fix based on @zimbatm comments in https://github.com/NixOS/nixpkgs/pull/76427

### CODE REVIEW

**Comment by eadwu** [comment]
Just a rough look so far but this looks like a `vscode-with-extensions` but just with mutable directories so that extensions can work without "heavy" patching instead of a `vscodeWithConfiguration` (which refers to the configuration file `settings.json` at first glance to me).

**Comment by countoren** [comment]
@eadwu I might not completely understand your comment(feel free to correct me) but in general:
vscode with extension installs the extensions in nix store, sometimes the extensions needs mutate files in the extension folder. base on the nix philosophy you don't want directories in the nix store to be altered. but you want to have to ability declaratively declare extensions that wont change or alter too. the problem here that vscode have in general 1 target extensions folder so in order to get the best of the both worlds(and to be able to use vscodeExts2nix which will get you the extensions that were chosen from vscode interface into a nix expression) you need to have an extensions folder that vscode could install new/update extensions into it.

File: pkgs/misc/vscode-extensions/vscodeEnv.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#use vscodeWithConfiguration and vscodeExts2nix to create vscode exetuable that when exits(vscode) will update the mutable extension file, which is imported when getting evaluated by nix.
+{ pkgs                            ? import <nixpkgs> {}
```
Comment (eadwu) [inline pkgs/misc/vscode-extensions/vscodeEnv.nix]: Don't really see the point of re-importing `nixpkgs` here, most of the things here are already given (besides `vscode-utils.extensionsFromVscodeMarketplace`) which could be easily added.

File: pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+, mutableExtensions               ? []
+, vscodeExtsFolderName            ? ".vscode-exts"
+}:
+let 
+  nixExtsDrvs = extensionsFromVscodeMarketplace nixExtensions;
+  mutExtsDrvs = extensionsFromVscodeMarketplace mutableExtensions;
+
+  #removed not defined extensions
+  rmExtensions =  lib.optionalString (nixExtensions++mutableExtensions != []) ''
+    find ${vscodeExtsFolderName} -mindepth 1 -maxdepth 1 ${lib.concatMapStringsSep " " (e : ''! -iname ${e.publisher}.${e.name}'') (nixExtensions++mutableExtensions)} -exec sudo rm -rf {} \;
```
Comment (eadwu) [inline pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix]: I don't really know how I feel about the fact `sudo` has to be used here.

File: pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  '';
+  #copy mutable extension out of the nix store
+  cpExtensions = ''
+    ${lib.concatMapStringsSep "\n" (e : ''ln -sfn ${e}/share/vscode/extensions/* ${vscodeExtsFolderName}/'') nixExtsDrvs}
+    ${lib.concatMapStringsSep "\n" (e : ''
+      cp -a ${e}/share/vscode/extensions/${e.vscodeExtUniqueId} ${vscodeExtsFolderName}/${e.vscodeExtUniqueId}-${(lib.findSingle (ext: ''${ext.publisher}.${ext.name}'' == e.vscodeExtUniqueId) "" "m" mutableExtensions ).version}
```
Comment (eadwu) [inline pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix]: `cp -a` keeps the directory permissions the same, so the extension directory wouldn't be mutable no? The extension directory permissions for me are 555.

File: pkgs/misc/vscode-extensions/vscodeEnv.nix (hunk 1 above)
Comment (countoren) [reply pkgs/misc/vscode-extensions/vscodeEnv.nix]: @eadwu sorry this was the wrong branch look at PR https://github.com/NixOS/nixpkgs/pull/76427 nixpkgs was removed there.

File: pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix (hunk 2 above)
Comment (countoren) [reply pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix]: yes I don't like it too, any better idea for syncing the extensions folder (we need to be consistent with the nix expression when vscode starts) ?

File: pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix (hunk 3 above)
Comment (countoren) [reply pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix]: for me on drawin it solved the permission problem on the csharp extension. but maybe it is not the same on a different OS. will changing permissions make sense?

**Comment by countoren** [comment]
outdated look at PR https://github.com/NixOS/nixpkgs/pull/76427

**Pull request closed without merging**